com.unity.perception/Runtime/Randomization/Scenarios/PerceptionScenario.cs

```diff
@@ -0,0 +1,94 @@
+
+namespace UnityEngine.Perception.Randomization.Scenarios
+{
+    /// <summary>
+    /// Derive this class to configure perception data capture while coordinating a scenario
+    /// </summary>
+    /// <typeparam name="T">The type of scenario constants to serialize</typeparam>
+    public abstract class PerceptionScenario<T> : Scenario<T> where T : ScenarioConstants, new()
+    {
+        /// <summary>
+        /// The guid used to identify this scenario's Iteration Metric Definition
+        /// </summary>
+        protected const string k_ScenarioIterationMetricDefinitionId = "DB1B258E-D1D0-41B6-8751-16F601A2E230";
+
+        /// <summary>
+        /// The metric definition used to report the current scenario iteration
+        /// </summary>
+        protected MetricDefinition m_IterationMetricDefinition;
+
+        /// <summary>
+        /// The scriptable render pipeline hook used to capture perception data skips the first frame of the simulation
+        /// when running locally, so this flag is used to track whether the first frame has been skipped yet.
+        /// </summary>
+        protected bool m_SkippedFirstFrame;
+
+        /// <inheritdoc/>
+        protected override bool isScenarioReadyToStart
+        {
+            get
+            {
+                if (!m_SkippedFirstFrame)
+                {
+                    m_SkippedFirstFrame = true;
+                    return false;
+                }
+                return true;
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void OnAwake()
+        {
+            m_IterationMetricDefinition = DatasetCapture.RegisterMetricDefinition(
+                "scenario_iteration", "Iteration information for dataset sequences",
+                Guid.Parse(k_ScenarioIterationMetricDefinitionId));
+        }
+
+        /// <inheritdoc/>
+        protected override void OnStart()
+        {
+            var randomSeedMetricDefinition = DatasetCapture.RegisterMetricDefinition(
+                "random-seed",
+                "The random seed used to initialize the random state of the simulation. Only triggered once per simulation.",
+                Guid.Parse("14adb394-46c0-47e8-a3f0-99e754483b76"));
+            DatasetCapture.ReportMetric(randomSeedMetricDefinition, new[] { genericConstants.randomSeed });
+        }
+
+        /// <inheritdoc/>
+        protected override void OnIterationStart()
+        {
+            DatasetCapture.StartNewSequence();
+            DatasetCapture.ReportMetric(m_IterationMetricDefinition, new[]
+            {
+                new IterationMetricData { iteration = currentIteration }
+            });
+        }
+
+        /// <inheritdoc/>
+        protected override void OnComplete()
+        {
+            DatasetCapture.ResetSimulation();
+            Manager.Instance.Shutdown();
+        }
+
+        /// <inheritdoc/>
+        protected override void OnIdle()
+        {
+            if (Manager.FinalUploadsDone)
```

Comment:
This should not be necessary. Calling Quit() will automatically call Manager.Instance.Shutdown() and should terminate synchronously.

Reply:
Ahh, I'll need to test this again then. In the past, Devin and I ran into a problem with how the final uploads took a few frames to complete before we could quit, or else the simulation would not save all the recorded data properly.

Reply:
Yep, this totally works. The scenario now quits within the OnComplete() hook